Pass the session's proxies property to request

junqfisica opened the pull request:
After waiting a long time for a fix on the side of the request that could fix pip install issues with proxy as already discussed in #9691. It seems that such changes as mentioned in psf/requests#5735 will take a while to happen, probably only on version 3 of the request. 

Therefore, I'm suggesting a change on the pip side as already proposed in https://github.com/pypa/pip/issues/9568#issuecomment-790516875 and https://github.com/psf/requests/pull/5735#issuecomment-939373868 

I think it's important that pip address this issue asap since the newer versions of pip are not working properly behind a proxy.

<!---
Thank you for your soon to be pull request. Before you submit this, please
double check to make sure that you've added a news file fragment. In pip we
generate our NEWS.rst from multiple news fragment files, and all pull requests
require either a news file fragment or a marker to indicate they don't require
one.

To read more about adding a news file fragment for your PR, please check out
our documentation at: https://pip.pypa.io/en/latest/development/contributing/#news-entries
-->

pfmoore commented:
This needs a test, to ensure that (a) it fixes the issue, and (b) we don't introduce regressions with future changes.

pradyunsg commented:
Do we have tests for any sort of proxy logic? If not, it might be fine to merge this as-is. 💁🏽

That said, there’s no version of this where having an actual test written is problematic, so it would 100% be a good idea to have one here.

pfmoore commented:
I don't know enough myself about proxy handling to be sure it's correct without a test, but I'm fine if someone else wants to merge it as is.

CrazyBoyFeng commented:
Simulating the behavior of a proxy is a complicated matter. If it is possible to import third-party libraries into the tests, the most efficient way is to create a real proxy server.

junqfisica commented:
@pfmoore Thank you for your clarification. 
I agree that some sort of server mocking would be better, but as already mentioned this is not an easy task. I would be glad to help if someone has an idea of how to do it, otherwise, you could just remove this test.
 
I'm sorry for the previous error, I just realized there was an indentation problem on the test committed since I just copy and paste my local changes directly on the raw file on Github. I will clone this branch so that doesn't happen again...sorry

uranusjr commented:
The change looks good to me.

Can you add `@pytest.mark.network` to the test to signify the test relies on network? It allows skipping the test in non-internet-connected setups. (I think that’s the marker to use; please search the codebase to ensure correct spelling.)

Also please add a news entry for this.

pfmoore commented:
Given the provided test will always fail on our CI, it's not much use IMO (anyone who can manually set a proxy to exercise the test can probably also set their proxy and just try pip out, after all...). My main intention in asking for a test was so that people *without* the ability to set up a proxy would be covered, because the test would run in CI to ensure this functionality still worked.

Question for the other @pypa/pip-committers - do we want to include this change without it being tested by our CI?

I'm ambivalent. The change is simple enough that asking for a test is making the whole PR turn into a much bigger exercise than it should be[^1]. But on the other hand, if we do ever switch to another network library like httpx, how will we know that proxy support still works if we have no test?

[^1]: And @junqfisica thanks for being patient in working with us on this.

uranusjr commented:
I think we should; if we don’t want to add anything non-testable, we should’ve not supported proxies to begin with 🙂

notatallshaw commented:
What about an integration test where a simple proxy is set up and then it tests that it can use it to say download?

I'm experimenting with it now, would you be interested with in a PR if I get it working okay?

pradyunsg commented:
I take it that https://github.com/pypa/pip/pull/10901 is the PR.

notatallshaw commented:
> I take it that #10901 is the PR.

Yes

junqfisica commented:
Hi everyone,

I have updated the test_network_session and added the `@pytest.mark.network` as @uranusjr suggested. I also added a --proxy keyword so users can run the test with a given proxy as suggested by @pfmoore.

Now the test works as follow: 

1. If the system has a proxy, the test tries to get it and it should pass.
2. If the system has no proxy the test also should pass as far as a proxy is not required by your network. 
3. Users can pass a proxy to the test using the --proxy keyword, if a valid proxy is passed then the test should pass.

Examples of how to run: 
 Should Pass if the system has a valid proxy or no proxy.
`pytest -s tests/unit/test_network_session.py::TestPipSession::test_proxy`
Should Pass.
`pytest -s tests/unit/test_network_session.py::TestPipSession::test_proxy --proxy valid_proxy` 
Should Fail.
`pytest -s tests/unit/test_network_session.py::TestPipSession::test_proxy --proxy invalid_proxy` 


@uranusjr I know you ask me to add a news entry, but I'm not sure how/what exactly should I add. I know there is a `NEWS.rst` file you want me to add a new entry there, right? If so, could give me some guidance on this I would appreciate it.

junqfisica commented:
> Given the provided test will always fail on our CI, it's not much use IMO (anyone who can manually set a proxy to exercise the test can probably also set their proxy and just try pip out, after all...). My main intention in asking for a test was so that people _without_ the ability to set up a proxy would be covered, because the test would run in CI to ensure this functionality still worked.
> 
> Question for the other @pypa/pip-committers - do we want to include this change without it being tested by our CI?
> 
> I'm ambivalent. The change is simple enough that asking for a test is making the whole PR turn into a much bigger exercise than it should be[1](#user-content-fn-1-9ec35cc377e20c7f46dfccaecd9123e8). But on the other hand, if we do ever switch to another network library like httpx, how will we know that proxy support still works if we have no test?
> 
> ## Footnotes
> 1. And @junqfisica thanks for being patient in working with us on this. [↩](#user-content-fnref-1-9ec35cc377e20c7f46dfccaecd9123e8)

It is my pleasure @pfmoore to help and give a small contribution to such an important package to the python community as pip.

pradyunsg commented:
> If so, could give me some guidance on this I would appreciate it.

Happy to! Please see https://pip.pypa.io/dev/news-entry-failure.

This is mentioned in the [CI failure for the news file check](https://github.com/pypa/pip/runs/5175453243?check_suite_focus=true#step:3:14) as well, although that hadn't been run when you asked your question. :)

junqfisica commented:
@pradyunsg Thank you.

I've added a News entry [9691.bugfix.rst](https://github.com/pypa/pip/pull/10680/commits/ea941234bbc943a969b312e56e4e21eca7763bbd#diff-c2b7e1afee5880e2168533a28bc2d79985beafac19c5926289facd5239ee461e) that makes reference to the issue https://github.com/pypa/pip/issues/9691. Hope that's it.

junqfisica commented:
Hi everyone, 

First I would like to thank you all for the support on this PR. 
Now that all checks have passed, is there any time estimative when this PR will be merged? :)
